Draft: Workflow improvements to run independently of BETY

yinghaoSunn opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
This draft Pull Request introduces changes to allow the LPJ-GUESS workflow to run independently of the BETY database. The main changes include:
- Updated database connection logic to handle missing BETY settings.
- Added fallback mechanisms for standalone workflow operation.

Feedback and further suggestions are welcome.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I agree that PEcAn Project may distribute my contribution under any or all of
	- the same license as the existing code,
	- and/or the BSD 3-clause license.
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

Commit:
Fix the problem where checking qsub status.

Commit:
Allow workflow to run independently of BETY database

infotroph commented on base/workflow/R/runModule.run.write.configs.R:
```diff
@@ -13,7 +13,6 @@ runModule.run.write.configs <- function(settings, overwrite = TRUE) {
     }
     return(PEcAn.settings::papply(settings, runModule.run.write.configs, overwrite = FALSE))
   } else if (PEcAn.settings::is.Settings(settings)) {
-    write <- settings$database$bety$write
```
Can you explain the intended effect of deleting this line? My first guess is maybe the idea was for `settings$database$bety$write` to no longer get consulted at `run.write.configs` time and instead to have the ensemble and get.parameter.samples steps check it directly? But then I'm not sure why the changes in `run.write.configs` were needed, so I think I'm off-track already.

I ask because this deletion is what's causing the current CI failure -- it leaves `write` undefined in the call to run.write.configs on line 29 below, which then means the references to `write` inside run.write.configs get interpreted as the base function `write()` instead of as a logical value. I can think of several possible fixes ranging from re-adding this line to renaming the argument so it doesn't conflict with a base function name, but without knowing more about the intention I don't know which approach to recommend.

yinghaoSunn replied:
Sorry for missing your comment earlier; I was busy with finals at the time. You’re absolutely right—I was trying to skip consulting `settings$database$bety$write`, but that shouldn’t have been done in `run.write.configs`. Your valuable comment has helped me since I was in the middle of troubleshooting the CI failure. Thanks!

infotroph replied:
In that case re-adding this line, together with my suggested change in run.write.configs above that @dlebauer already applied for us, should be what you need to get CI passing 🤞

infotroph commented on base/workflow/R/run.write.configs.R:
```diff
@@ -25,14 +25,20 @@
-  })
+  ## Skip database connection if settings$database is NULL
+  if (write == FALSE) {
+    PEcAn.logger::logger.info("Database connection skipped: No database settings provided.")
```
```suggestion
  if (!isTRUE(write)) {
    PEcAn.logger::logger.info("Not writing this run to database, so database connection skipped")
```

infotroph committed:
Merge branch 'develop' into develop

dlebauer committed:
Update base/workflow/R/run.write.configs.R

Co-authored-by: Chris Black <chris@ckblack.org>

dlebauer commented on modules/uncertainty/R/ensemble.R:
```diff
@@ -236,6 +237,7 @@ write.ensemble.configs <- function(defaults, ensemble.samples, settings, model,
       PEcAn.logger::logger.warn("We were not able to successfully establish a connection with Bety ")
```
```suggestion
      PEcAn.logger::logger.warn("We were not able to successfully establish a connection with BETYdb ")
```

dlebauer committed:
Update modules/uncertainty/R/ensemble.R

dlebauer commented on modules/uncertainty/R/get.parameter.samples.R:
```diff
@@ -202,7 +206,7 @@ get.parameter.samples <- function(settings,
     }
   }
-  
+  #ensemble.samples <- NULL
```
why add this commented out line?

yinghaoSunn replied:
Sorry, it was part of a previous attempt. It should be deleted, but I forgot.

yinghaoSunn committed:
Merge branch 'develop' into develop

yinghaoSunn committed:
Remove unused commented-out line

yinghaoSunn committed:
Update runModule.run.write.configs.R

Re-add the `write <- settings$database$bety$write` that was mistakenly deleted earlier.

infotroph commented on modules/uncertainty/R/ensemble.R:
```diff
@@ -222,6 +222,7 @@ write.ensemble.configs <- function(defaults, ensemble.samples, settings, model,
   }
   
   # See if we need to write to DB
+  if (!is.null(settings$database) && !is.null(settings$database$bety) && !is.null(settings$database$bety$write)) {
```
If a list entry exists we know all its parents exist too, so I think this can be safely simplified to just 
```suggestion
  if (!is.null(settings$database$bety$write)) {
```

infotroph commented on modules/uncertainty/R/get.parameter.samples.R:
```diff
@@ -202,7 +206,6 @@ get.parameter.samples <- function(settings,
                                                env.samples, ens.sample.method, param.names)
     }
   }
-  
```
nit: I suggest re-adding the empty line here -- It seems to be accepted R style  to have whitespace before the final (and thus implicitly returned) statement of a function.

yinghaoSunn replied:
Thanks a lot！

infotroph reviewed (approved):
Thanks! this looks good and I'm happily building on it already for DB-free SIPNET runs.

yinghaoSunn committed:
Update modules/uncertainty/R/ensemble.R

Co-authored-by: Chris Black <chris@ckblack.org>

yinghaoSunn committed:
Update get.parameter.samples.R

Re-add an empty line before the final statement in the get.parameter.samples function to align with common R style guidelines.

yinghaoSunn committed:
Merge branch 'PecanProject:develop' into develop

yinghaoSunn committed:
Update run.write.configs.R

Updated the conditional logic to handle cases where database writes are disabled (`write = FALSE`) or the database settings are `NULL`.

infotroph commented on base/workflow/R/run.write.configs.R:
```diff
@@ -25,11 +25,14 @@
-  } else {
+  } else if(is.null(settings$database)) {
+    PEcAn.logger::logger.info("There's no database connection, so database connection skipped")
+    con <- NULL  # Set con to NULL to avoid errors in subsequent code
```
I'm only about 60% sure of this, but I'm inclined to think this should be an error -- we only hit this condition when `write = TRUE`, which seems like reasonably strong evidence the user really does want to use the database. Thoughts?

mdietze replied:
Correct. I discussed this with @yinghaoSunn last week as the most likely source of his GH Actions failure and we discussed him changing the first `if` case to be `!isTRUE(write) && is.null(settings$database)` and then within the `else` to have the second `if` throwing an error (not a warning) with an informative error message if `is.null(settings$database)` because that means you want to write but dont have a db connection. The second else would then have the tryCatch to open the connection.  It's important that the very first `if` check both write and db being null because GH actions is failing in the case where write == FALSE but the db connection is still needed to READ from the db to get information that's not being provided in the pecan.xml (which presumably IS provided by the user when they don't provide settings$database)

yinghaoSunn replied:
Thank you for pointing this out and for the detailed explanation! It was an oversight on my part to skip over this scenario without throwing an informative error. I’ve updated the logic to handle the condition where `write == TRUE` but `settings$database == NULL`.

yinghaoSunn committed:
Update run.write.configs.R

Added a new condition to explicitly stop execution if `write = TRUE` but `settings$database` is `NULL`, providing a clear error message to guide the user.

yinghaoSunn committed:
Merge branch 'develop' into develop

yinghaoSunn committed:
Update get.parameter.samples.R

Just a try.

infotroph commented on modules/uncertainty/R/ensemble.R:
```diff
@@ -222,6 +222,7 @@ write.ensemble.configs <- function(defaults, ensemble.samples, settings, model,
   }
   
   # See if we need to write to DB
+  if (!is.null(settings$database$bety$write)) {
   write.to.db <- as.logical(settings$database$bety$write)
```
Should have the same effect but make it a bit clearer what's happening (I think the fact the existing function completely ignored the `write.to.db` in the function args was a bug).
Might also be worth updating the Roxygen entry on line 198 to something like `@param write.to.db logical: Record this run in BETY? Note that if settings$database$bety$write is set, it will override this`
```suggestion
  if (!is.null(settings$database$bety$write)) {
    # specifying `write` in settings overrides write.to.db in fn args
    write.to.db <- as.logical(settings$database$bety$write)
  }
```

yinghaoSunn replied:
I agree. It will help understand the parameter priority.

infotroph commented on modules/uncertainty/R/ensemble.R:
```diff
@@ -233,9 +234,10 @@ write.ensemble.configs <- function(defaults, ensemble.samples, settings, model,
+      PEcAn.logger::logger.warn("We were not able to successfully establish a connection with BETYdb ")
     }
   }
+  }
```
Only accept this one if also accepting my change above ;)
```suggestion
```

infotroph commented on modules/uncertainty/R/get.parameter.samples.R:
```diff
@@ -202,7 +204,7 @@ get.parameter.samples <- function(settings,
     }
   }
-  
+                                
```
```suggestion

```

infotroph committed:
Update modules/uncertainty/R/get.parameter.samples.R

infotroph commented on modules/uncertainty/R/get.parameter.samples.R:
```diff
@@ -202,7 +204,6 @@ get.parameter.samples <- function(settings,
     }
   }
-  
   save(ensemble.samples, trait.samples, sa.samples, runs.samples, env.samples, 
```
```suggestion

  save(ensemble.samples, trait.samples, sa.samples, runs.samples, env.samples, 
```

infotroph committed:
Update modules/uncertainty/R/get.parameter.samples.R

yinghaoSunn committed:
Update modules/uncertainty/R/ensemble.R

Clarified the logic for assigning `write.to.db`, ensuring it is overridden by `settings$database$bety$write` when provided.

Co-authored-by: Chris Black <chris@ckblack.org>

yinghaoSunn committed:
Update modules/uncertainty/R/ensemble.R

Clarified the logic for assigning `write.to.db`, ensuring it is overridden by `settings$database$bety$write` when provided

Co-authored-by: Chris Black <chris@ckblack.org>

yinghaoSunn committed:
Update get.parameter.samples.R

This info is not needed.

infotroph commented on base/workflow/R/run.write.configs.R:
```diff
@@ -25,14 +25,25 @@
+    PEcAn.logger::logger.info("Not writing this run to database, so database connection skipped")
+    con <- NULL  # Set con to NULL to avoid errors in subsequent code
+  } else if(is.null(settings$database)) {
+  PEcAn.logger::logger.error(
+    "Database is NULL but writing is enabled. Provide valid database settings in pecan.xml."
+  )
+  stop("Database connection required but settings$database is NULL.")
```
```suggestion
    PEcAn.logger::logger.error(
      "Database is NULL but writing is enabled. Provide valid database settings in pecan.xml."
    )
    stop("Database connection required but settings$database is NULL.")
```

infotroph committed:
Update base/workflow/R/run.write.configs.R

infotroph commented on base/workflow/R/start_model_runs.R:
```diff
@@ -170,9 +172,10 @@ start_model_runs <- function(settings, write = TRUE, stop.on.error = TRUE) {
+      if (write){
+        # write finished time to database
+        PEcAn.DB::stamp_finished(con = dbcon, run = run)
+        }
```
```suggestion
      }
```

infotroph committed:
Update base/workflow/R/start_model_runs.R

infotroph commented:
@yinghaoSunn Thank you for the work on this! I would be ready to merge it now, but since it's still marked as a draft I wanted to check first if there are other changes you're planning to add.

yinghaoSunn commented:
> @yinghaoSunn Thank you for the work on this! I would be ready to merge it now, but since it's still marked as a draft I wanted to check first if there are other changes you're planning to add.

Thank you for checking! I don’t have anything else to add. I really appreciate the support and guidance from you, David, and Mike throughout this process. This is my first Pull Request, and I’m thrilled to contribute to the community.
